src/classes/TG_ValidateData_TEST.cls

```diff
@@ -444,38 +446,226 @@ private class TG_ValidateData_TEST {
         if (UTIL_Describe_API.isValidField(UTIL_Namespace.StrTokenNSPrefix('Course_Enrollment__c'),'RecordTypeId')) { 
 
             Contact student = [SELECT id FROM Contact LIMIT 1];
-        
             Course_Offering__c courseOffering = [SELECT id FROM Course_Offering__c LIMIT 1];
-        
             Course_Enrollment__c courseEnrollment = new Course_Enrollment__c(Course_Offering__c = courseOffering.Id, contact__c = student.id);
             courseEnrollment.put('RecordTypeId', UTIL_Describe_API.getStudentConnectionRecType());
             insert courseEnrollment;
-        
             Course__c course = [SELECT Id FROM Course__c LIMIT 1];
-         
             Term__c term = [SELECT Id FROM Term__C LIMIT 1];
-         
             Course_Offering__c courseOfferingNew = new Course_Offering__c(Course__c = course.id, Term__c = term.id, Name = 'ENG-201');
             insert courseOfferingNew;
-         
             Course_Enrollment__c courseEnrollmentNew = new Course_Enrollment__c(Course_Offering__c = courseOfferingNew.Id, contact__c = student.id);
             courseEnrollmentNew.put('RecordTypeId', UTIL_Describe_API.getStudentConnectionRecType());
             insert courseEnrollmentNew;
-         
+            Test.startTest();
             List<Term_Grade__c> termGrades = new List<Term_Grade__c>();
             Term_Grade__c termGrade1 = new Term_Grade__c(Course_Offering__c = courseOffering.id, contact__c = Student.id);
             termGrades.add(termGrade1);
-         
             Term_Grade__c termGrade2 = new Term_Grade__c(Course_Offering__c = courseOfferingNew.id, contact__c = Student.id);
             termGrades.add(termGrade2);
-         
             insert termGrades;
-        
+            Test.stopTest();
             Term_Grade__c updatedTermGrade = [SELECT Course_Connection__c FROM Term_Grade__c WHERE Course_Offering__c =:courseOffering.id LIMIT 1];
              
             System.assertEquals(updatedTermGrade.Course_Connection__c, courseEnrollment.id);  
            
         }
 
-     }   
+     }  
+    
+    // Test TermGradeValidator class for valid fields populated
```

Comment:
@skosgi1 -- The beautiful thing about using the validator is that we actually shouldn't need any of the setup data. We can quite literally test if the hasValidFields() method is working by passing in a term grade with fake ids. We don't have to persist, etc. Your other unit tests are testing whether or not things are actually working as expected but the validator itself doesn't *need* to know about any of that other data because all it's concerned with is whether or not the term grade object has data in the fields we care about. It only needs an instance of the object and doesn't even have to be a persisted instance since you're checking that functionality in your other tests.
You can imagine if we started doing more things like this that or code would get simpler. Classes and methods only needing to deal with the problem at hand rather than having to have the bigger picture. We still want tests for the bigger picture (and you've done that). But the tests for the validator itself can be simplified:

Term_Grade__c tg = new Term_Grade__c(Course_Connection__c = 'fake', Contact__c = 'morefake', Course_Offering__c = 'reallyfake');
TermGradeValidator tgv = new TermGradeValidator(tg);
System.assertEquals(true, tgv.hasValidFields());

And similar tests to trigger and test for your various errors. Also for the "false" ones, we may want to assert that the "getError()" method is returning the Label we expect...

Reply:
Not sure if I am doing it right but when I do that I get the error "System.StringException: Invalid id: fake"

Reply:
are you trying to insert the record? If so, don't :) Strange...I'll take a look. It was working for me yesterday but maybe I was doing something silly.

Reply:
@skosgi1 my bad -- leave it for now -- really strange and I wanna continue to look at it, but this time it gave me the error. Sorry about that, in my testing yesterday it was working but maybe i didn't save or something :/

Reply:
I tried today and it didn't work so I plugged in actual ids, not sure if I messed something up..

Reply:
Nope you were correct. I forgot last night I was using a quick utility to fake ids and didn't include that in what I sent. I will get that setup tomorrow because its helpful to have. Essentially its just a bit of code that takes a Schema.SObjectType as an argument and can generate a bogus Id for it purely for testing. :)
I'll update tomorrow. But thanks for calling that out and we can leave it as is for now.